lib/Conversion/SeqToSV/SeqToSV.cpp

```diff
@@ -190,8 +191,100 @@ class ClockMuxLowering : public OpConversionPattern<ClockMuxOp> {
     return success();
   }
 };
+
+/// Map `seq.clock` to `i1`.
+struct SeqToSVTypeConverter : public TypeConverter {
+  SeqToSVTypeConverter() {
+    addConversion([](Type type) { return type; });
+    addConversion([](seq::ClockType type) {
+      return IntegerType::get(type.getContext(), 1);
+    });
+  }
+};
+
+/// Rewrite clock ports on instances.
+class InstanceLowering : public OpConversionPattern<hw::InstanceOp> {
+public:
+  using OpConversionPattern<hw::InstanceOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(hw::InstanceOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+
+    SmallVector<Type, 1> results;
+    if (failed(typeConverter->convertTypes(op.getResultTypes(), results)))
+      return failure();
+
+    rewriter.replaceOpWithNewOp<hw::InstanceOp>(
+        op, results, op.getInstanceNameAttr(), op.getModuleNameAttr(),
+        adaptor.getInputs(), op.getArgNamesAttr(), op.getResultNamesAttr(),
+        op.getParameters(), op.getInnerSymAttr());
+
+    return success();
+  }
+};
+
+/// Rewrite the operands of users of clock types.
+template <typename T>
+class ClockUserLowering : public OpConversionPattern<T> {
+public:
+  using OpConversionPattern<T>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(T op, typename T::Adaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    // For an op which only uses clocks, rewrite the operands.
+    rewriter.updateRootInPlace(op,
+                               [&] { op->setOperands(adaptor.getOperands()); });
+    return success();
+  }
+};
+
+/// Eliminate no-op clock casts.
+template <typename T>
+class ClockCastLowering : public OpConversionPattern<T> {
+public:
+  using OpConversionPattern<T>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(T op, typename T::Adaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    rewriter.replaceOp(op, adaptor.getInput());
+    return success();
+  }
+};
+
+/// Lower a `hw.wire` which produces clock types.
+class ClockWireLowering : public OpConversionPattern<hw::WireOp> {
```

Comment:
Why is hw.wire different from any other op? Wouldn't this be handled correctly by TypeConversionPattern? (I must be missing something.)

Reply:
Removed.